File: 프로그래머스/2/42885. 구명보트/구명보트.js
class Deque {
    constructor() {
        this.deque = new Map();
        this.head = 0;
        this.tail = 0;
    }
    
    get size() {
        return this.tail - this.head;
    }
    
    isEmpty() {
        return this.size === 0;
    }
    
    push(element) {
        this.deque.set(this.tail, element);
        this.tail += 1;
    }
    
    unshift(element) {
        this.head -= 1;
        this.deque.set(this.head, element);
    }
    
    pop() {
        if (this.isEmpty()) return null;
        this.tail -= 1;
        const lastElement = this.deque.get(this.tail);
        this.deque.delete(this.tail);
        return lastElement;
    }
    
    shift() {
        if (this.isEmpty()) return null;
        const firstElement = this.deque.get(this.head);
        this.deque.delete(this.head);
        this.head += 1;
        return firstElement;
    }
    
    get first() {
        return this.deque.get(this.head);
    }
    
    get last() {
        return this.deque.get(this.tail - 1);
    }
}

function solution(people, limit) {
    const sortedPeople = people.sort((a, b) => b - a);
    const deque = new Deque();
    sortedPeople.forEach((person) => {
        deque.push(person);
    });
    
    let boatCount = 0;
    while (!deque.isEmpty()) {
        if (deque.size >= 2 && (deque.first + deque.last <= limit)) {
            deque.shift();
            deque.pop();
        } else if (deque.first <= limit) {
            deque.shift();
        } else {
            throw deque;
        }
        boatCount += 1;
    }
    return boatCount;
}

// 50_000 * 50_000 = 25 * 10^8 = 2.5 * 10^9
// 1_000_000_000

// 정렬 후 deque에 넣기
// 가장 큰 수 꺼내서 확인 -> 
//    1. 가장 작은 수와 합 <= limit -> 둘 다 빼버리기
//    2. 가장 작은 수와 합 > limit > 하나만 넣기

File: 프로그래머스/2/42885. 구명보트/구명보트.py
from collections import deque

def solution(people, limit):
    
    sorted_people = sorted(people, reverse = True)
    elements = deque(sorted_people)
    
    count = 0
    while len(elements) > 0:
        first = elements.popleft()
        if len(elements) > 0 and first + elements[len(elements) - 1] <= limit:
            second = elements.pop()
        count += 1
        
    return count
        
    
#     partners = set()
#     count = 0
#     for index, person_weight in enumerate(sorted_people):
#         if index in partners:
#             continue
#         rest = limit - person_weight
#         target = bisect_left(sorted_people, rest)
#         print(person_weight, target, sorted_people[target], rest)
#         count += 1
    


# 25 * 10 ** 8 => 25억: O(n^2) 시간초과

# 50 50 70 80

# arr 정렬  nlogn (n <= 50_000)
# while arr is not empty:
#   element = arr.deque
#   rest = capacity - element
#   find biggest i such that i <= rest
#   문제: arr에 중복값이 있을 수 있음

# 내림차순으로 정렬 후,
# 각 요소에 대하 popleft할 때 가장 작은 것도 같이 담을 수 있으면 담기
# exchange argument
# a0, ai
# ai가 들어가는 곳에 bi가 들어간다고 하면, ai와 bi를 교체 가능






